### ISSUE TITLE
[ [iOS] [4.0.0-rc04] Setting repeat mode to RepeatMode.Track causes playback to pause after completing a track

### ISSUE THREAD

**Issue by TobiDevloft** [doublesymmetry/react-native-track-player#1995]
**Describe the Bug**
When setting repeat mode to `RepeatMode.Track` on iOS (library version 4.0.0-rc04), playback gets paused after the track is completed. The progress is correctly reset. No `PlaybackActiveTrackChanged` event is fired.

On android, `RepeatMode.Track` works as expected. The track restarts after completing, and a `PlaybackActiveTrackChanged` event is fired.

**Steps To Reproduce**

On iOS, Set repeat mode to `RepeatMode.Track` by calling `await TrackPlayer.setRepeatMode(RepeatMode.Track)`. Then finish a track.

**Code To Reproduce**
```
export const SetupService = async () => {
     await TrackPlayer.setupPlayer();
     await TrackPlayer.setRepeatMode(RepeatMode.Track)
}

```

**Replicable on Example App?**
I created a[ fork of this project](https://github.com/TobiDevloft/react-native-track-player) where repeat mode is set to `RepeatMode.Track` in the example app, [in this commit](https://github.com/TobiDevloft/react-native-track-player/commit/b0941b57a9ad50a1d67e0a06f3ad0596c448d39a).

**Paste the results of `npx react-native info`:** 

```
System:
    OS: macOS 13.2.1
    CPU: (12) arm64 Apple M2 Pro
    Memory: 104.02 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.10.0 - ~/.nvm/versions/node/v16.10.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.10.0/bin/yarn
    npm: 7.24.0 - ~/.nvm/versions/node/v16.10.0/bin/npm
    Watchman: 2023.03.06.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.12.0 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    Android SDK:
      Android NDK: 25.2.9519653
  IDEs:
    Android Studio: Not Found
    Xcode: 14.3/14E222b - /usr/bin/xcodebuild
  Languages:
    Java: 16.0.2 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0
    react-native: 0.71.4 => 0.71.4
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

**Paste the exact `react-native-track-player` version you are using**

`4.0.0-rc04`

**Real device? Or simulator?**

Real device

What OS are you running?

iOS 16.4

**How I can Help**

I'm pretty sure this is an issue with `SwiftAudioEx`. I'd love to help, but I don't know Swift and how to integrat work on `SwiftAudioEx` into this project during development, and my experience with native development is generally very limited. So I'm not sure if I can be of help here unfortunately.

**Comment by kirillzyusko** [comment]
I also can confirm the issue is present in latest `4.0.0-rc04` release.

**Comment by 2017398956** [comment]
me too

**Comment by 2017398956** [comment]
This is how I temporarily solved the problem：
`if (Platform.OS === 'ios') {
			await this.playSoundOnce(url, stop => {
				if (stop) {
					this.playSoundOnce(url);
				}
			});
		} else {
			await this.playSoundWithMode(url, RepeatMode.Queue);
		}`

**Comment by 2017398956** [comment]
This is how I temporarily solved the problem

`if (Platform.OS === 'ios') {
			let callback = (stop: boolean) => {
				if (stop) {
					this.playSoundOnce(url, callback);
				}
			};
			await this.playSoundOnce(url, callback);
		} else {
			await this.playSoundWithMode(url, RepeatMode.Queue);
		}`

**Comment by TobiDevloft** [comment]
I tried it again in `4.4.0-rc05`. Now the playback doesn't pause after completing the track, but no event is fired when the track restarts again. This unfortunately makes it next to impossible to build repeat mode functionality where a track is played n-times :/ Is this maybe just an issue on my end? Or something you experienced as well @jspizziri ?

**Comment by puckey** [comment]
Since the active track has not changed, why would you expect `PlaybackActiveTrackChanged` to fire? What happens on android, does it fire the event there?

**Comment by TobiDevloft** [comment]
I agree with you, but unfortunately I experience a few unintended side effects with repeat mode enabled.

`PlaybackActiveTrackChanged` is called in repeat mode on Android, but not in iOS. In 3.2.0 and before `PlaybackTrackChanged` was called for both OS every time the track started. `PlaybackTrackChanged`, though deprecated, is now also not called on iOS but still on Android.

What I just now noticed is that `PlaybackQueueEnded` is called with the last position of the track if repeat mode is not enabled. If repeat mode is enabled, it is called with position 0.

I currently work around both of these issues by listening to the player progress updates and comparing them with each other, which allows me to detect repeats. This is not ideal, but I didn't see another way. Let me know if I can provide further information!

**Comment by puckey** [comment]
> Thank you for responding & I agree with you @puckey! Unfortunately I experience a few unintended side effects with repeat mode enabled.
> 
> `PlaybackActiveTrackChanged` is called in repeat mode on Android when a track is restarted, but not in iOS. In 3.2.0 and before `PlaybackTrackChanged` was called for both OS every time the track started. `PlaybackTrackChanged`, though deprecated, is now also not called on iOS but still on Android.

Could you make an issue for the deprecated PlaybackTrackChanged not being called on ios?

I will check the Android implementation of the PlaybackActiveTrackChanged event, to see if we might change it to not be emitted when the same track repeats. Otherwise I will change the ios implementation to follow the android one.
 
> What I just now noticed is that `PlaybackQueueEnded` is fired with the last position of the track if repeat mode is not enabled. If repeat mode is enabled, it's fired with position 0.

 Could you make a separate issue for this?

> I currently work around both of these issues by listening to the player progress updates and comparing them with each other, which allows me to detect repeats. This is not ideal, but I didn't see another way. Let me know if I can provide further information!

What exactly are you doing that you need this workaround for?

**Comment by puckey** [comment]
(Good chance that the playback queue ended event issue is fixed by https://github.com/doublesymmetry/react-native-track-player/pull/2058)

**Comment by TobiDevloft** [comment]
I created https://github.com/doublesymmetry/react-native-track-player/issues/2059 and https://github.com/doublesymmetry/react-native-track-player/issues/2060

Thanks for your help, @puckey!

**Comment by TobiDevloft** [comment]
@puckey My assumption that this was fixed in` 4.0.0-rc5` was wrong unfortunately. This issue still exists in the current version, though it does not appear on an emulator - only on a real device. 

Should I close this in favor of https://github.com/doublesymmetry/react-native-track-player/issues/2070, or leave this open because it was about `4.0.0-rc04`?
